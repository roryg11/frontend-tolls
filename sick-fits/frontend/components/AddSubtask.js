import React, { Component } from 'react';
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import styled from "styled-components";
import Form from "./styles/Form";
import Error from "./ErrorMessage";
import SickButton from "./styles/SickButton";
import { GOAL_TASK_QUERY } from "./Task";

const CREATE_SUBTASK_MUTATION = gql`
    mutation CREATE_SUBTASK_MUTATION (
        $name: String!
        $description: String
        $taskId: ID!
    ) {
        createSubTask(
            name: $name
            description: $description
            taskId: $taskId
        ) {
            id
            name
            description
        }
    }
`

const FormStyles = styled.div`
    padding: 20px;
`

class AddSubtask extends Component {
    state = {
        taskId: this.props.taskId,
        name: "",
        description: ""
    }

    onChange = (e) => {
        const {name, value} = e.target; 
        this.setState({[name]: value});
    }
    render() {
        // will need to add refetch queries and animate the appearance of the subtask
        return (
            <Mutation mutation={CREATE_SUBTASK_MUTATION} variables={this.state} refetchQueries={[{query: GOAL_TASK_QUERY, variables: {id: this.props.taskId}}]}>
                { (addSubtask, {loading, error}) => {
                    if(loading) return <p>Loading...</p>;
                    if(error) return <Error error={error}/>;
                    return (
                        <FormStyles>
                            <Form onSubmit={
                            (e)=>{
                                e.preventDefault();
                                addSubtask();
                            }
                            }>
                                <label htmlFor="name">
                                        Name
                                        <input type="text"
                                        id="name"
                                        name="name"
                                        placeholder="Name"
                                        required
                                        value={this.state.name}
                                        onChange={this.onChange}/>
                                    </label>
                                    <label htmlFor="description">
                                        Description
                                        <input type="text"
                                        id="description"
                                        name="description"
                                        placeholder="How does this help achieve your milestone?"
                                        required
                                        value={this.state.description}
                                        onChange={this.onChange}/>
                                    </label>
                                    <SickButton type="Submit">Add Subtask</SickButton>
                            </Form>
                        </FormStyles>
                        
                    )
                }}
            </Mutation>
        );
    }
}

export default AddSubtask;