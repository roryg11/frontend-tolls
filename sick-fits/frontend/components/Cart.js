import {Query, Mutation} from "react-apollo";
import gql from "graphql-tag";
import CartStyles from "./styles/CartStyles"
import CloseButton from "./styles/CloseButton"
import SickButton from "./styles/SickButton"
import Supreme from "./styles/Supreme"

const LOCAL_STATE_QUERY = gql`
    query {
        cartOpen @client
    }
`

const TOGGLE_CART_MUTATION = gql`
    mutation {
        toggleCart @client
    }
`

const Cart = props => (
    <Mutation mutation={TOGGLE_CART_MUTATION}>
        { (toggleCart)=>(
            <Query query={LOCAL_STATE_QUERY}>
                { ({data}) => (
                    <CartStyles open={data.cartOpen}>
                        <header>
                            <CloseButton title="close" onClick={toggleCart}>&times;</CloseButton>
                            <Supreme>Your Cart</Supreme>
                            <p>You have __ items in your cart</p>
                        </header>
                        <footer>
                            <p>$10.10</p>
                            <SickButton>Checkout</SickButton>
                        </footer>
                    </CartStyles>
                )}
            </Query>
        )}
    </Mutation>
)

export default Cart; 
export { LOCAL_STATE_QUERY };
export { TOGGLE_CART_MUTATION }; 