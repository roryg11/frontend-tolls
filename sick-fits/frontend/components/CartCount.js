import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { TransitionGroup, CSSTransition} from "react-transition-group";

const AnimationStyles = styled.span`
    position: relative;
    .count {
        display: block;
        position: relative;
        transition: all 0.4s;
        backface-visibility: hidden;
    }
    /* initial state of entered dot flipped on back */
    /* one turn is 360 degrees */
    .count-enter{
        transform: rotateX(0.5turn);
    }
    .count-enter-active {
        transform: rotateX(0);
    }

    .count-exit {
        top: 0;
        position: absolute;
        transform: rotateX(0);
    }

    .count-exit-active {
        transform: rotateX(0.5turn);
    }
`;


// font-feature-settings and font-variant-numeric help wiht
// the fact that numbers have different widths -- this maintains the same width
const Dot = styled.div`
    background-color: ${props => props.theme.red};
    color: white;
    border-radius: 50%;
    padding: 0.5rem;
    line-height: 2rem;
    width: 3rem;
    margin-left: 1rem;
    font-weight: 100;
    font-feature-settings: tnum;
    font-variant-numeric: tabular-nums;
`

const CartCount = ({count}) => (
    <AnimationStyles>
        <TransitionGroup>
            <CSSTransition 
                unmountOnExit 
                className="count" 
                classNames="count" 
                key={count} 
                timeout={{enter: 400, exit: 400}}>
                <Dot>{count}</Dot>
            </CSSTransition>
        </TransitionGroup>
    </AnimationStyles>
    
)

export default CartCount;