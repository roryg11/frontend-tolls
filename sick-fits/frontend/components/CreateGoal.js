import React, { Component } from 'react';
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import Router from "next/router";
import styled from "styled-components";
import Error from "./ErrorMessage";
import Form from "./styles/Form";
import SickButton from "./styles/SickButton";

const MEASUREMENTS = [
    "FREQUENCY",
    "LIST"
]; 

const CREATE_GOAL_MUTATION = gql`
    mutation CREATE_GOAL_MUTATION(
        $name: String!
        $description: String!
        $measurement: Measurement
    ) {
        createGoal(
            name: $name,
            description: $description,
            measurement: $measurement
        ) {
            id
            name
        }
    }
`

// Future to-do, add inspiration images etc

class CreateGoal extends Component {
    state = {
        name: "",
        description: "",
        measurement: MEASUREMENTS.frequency
    }
    onChange = (e) => {
        const { name, value } = e.target; 
        this.setState({[name]: value}); 
    }
    render(){
        return(
            <Mutation mutation={CREATE_GOAL_MUTATION} variables={this.state}>
                { (createGoal, {loading, error}) => {
                        if(error) return <Error error={error}/>
                        if(loading) return <p>Loading...</p>
                        return (
                            <Form onSubmit={ async (e) => {
                                e.preventDefault();

                                const res = await createGoal();
                                Router.push({
                                    pathname: "/goal", 
                                    query: {id: res.data.createGoal.id}
                                });
                            }
                            }>
                                 <label htmlFor="name">
                                    Name
                                    <input type="text"
                                    id="name"
                                    name="name"
                                    placeholder="Name"
                                    required
                                    value={this.state.name}
                                    onChange={this.onChange}/>
                                </label>
                                <label htmlFor="description">
                                    Description
                                    <input type="text"
                                    id="description"
                                    name="description"
                                    placeholder="Why is this your goal?"
                                    required
                                    value={this.state.description}
                                    onChange={this.onChange}/>
                                </label>
                                <label htmlFor="measurement">
                                    Measurement - how will you measure your success?
                                    <select onChange={this.onChange} name="measurement" defaultValue={MEASUREMENTS.frequency}>
                                        {
                                            MEASUREMENTS.map((measurement)=>{
                                                return <option key={measurement} value={measurement}>{measurement}</option>;    
                                            })
                                        }
                                    </select>
                                </label>
                                <SickButton type="Submit">Onward!</SickButton>
                            </Form>
                        )
                    }
                }
            </Mutation>
        )
    }
}

export default CreateGoal;
export {MEASUREMENTS}; 

