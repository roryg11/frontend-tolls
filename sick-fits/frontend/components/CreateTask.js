import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from "graphql-tag";
import Router from "next/router";
import styled from "styled-components";
import Form from "./styles/Form";

const CREATE_TASK_MUTATION = gql`
    mutation CREATE_TASK_MUTATION (
        $name: String!
        $description: String
        $goalId: ID!
    ) {
        createTask(
            name: $name
            description: $description
            goalId: $goalId
        ) {
            id
            name
            description
        }
    }
`


class CreateTask extends Component {
    state = {
        name: "",
        description: "",
        goalId: this.props.goalId
    }

    handleChange = (e) => {
        const { name, value } = e.target; 
        this.setState({[name]: value});
    }

    render() {
        return (
            <Mutation mutation={CREATE_TASK_MUTATION} variables={this.state}>
                { (createTask, {loading, error})=>{
                        return (<Form onSubmit={
                                        async e=> {
                                            // stop form from submitting
                                            e.preventDefault();
                                            //call the mutation
                                            const res = await createTask();
                                            // change to the single item page
                                            Router.push({
                                                pathname: "/milestone", 
                                                query: {id: res.data.createTask.id}
                                            });
                                        }
                            }>
                            <fieldset disabled={loading} aria-busy={loading}>
                                <label htmlFor="name">
                                    Name
                                    <input type="text"
                                    id="name"
                                    name="name"
                                    placeholder="name"
                                    required
                                    value={this.state.title}
                                    onChange={this.handleChange}/>
                                </label>
                                <label htmlFor="description">
                                    Description
                                    <textarea
                                    id="description"
                                    name="description"
                                    placeholder="Enter a description"
                                    required
                                    value={this.state.description}
                                    onChange={this.handleChange}/>
                                </label>
                                <button type="Submit">Submit</button>
                            </fieldset>
                        </Form>)
                    }
                }
            </Mutation>
        );
    }
}

export default CreateTask;