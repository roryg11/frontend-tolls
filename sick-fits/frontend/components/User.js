import gql from "graphql-tag";
import { Query } from "react-apollo";
import propTypes from "prop-types";

const CURRENT_USER_QUERY = gql`
    query {
        me{
            id
            name
            email
            permissions
            cart {
                id
                quantity
                item {
                    id
                    title
                    description
                    image
                    price
                }
            }
        }
    }
`

const User = props => (
    <Query query={CURRENT_USER_QUERY}>
        { payload => props.children(payload)}
    </Query>
)

User.propTypes = {
    children: propTypes.func.isRequired,
}

export default User;
export {CURRENT_USER_QUERY};