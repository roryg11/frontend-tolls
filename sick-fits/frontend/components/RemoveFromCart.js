import React, { Component } from 'react';
import PropTypes from "prop-types";
import styled from "styled-components";
import gql from "graphql-tag";
import { Mutation } from "react-apollo";
import { CURRENT_USER_QUERY } from "./User";

const REMOVE_FROM_CART_MUTATION = gql`
    mutation REMOVE_FROM_CART_MUTATION($id: ID!){
        removeFromCart(id: $id){
            id
        }
    }
`

const BigButton = styled.button`
    font-size: 3rem;
    background: none;
    border: 0;
    &:hover {
        color: ${ props => props.theme.red};
        cursor: pointer;
    }
`;

class RemoveFromCart extends Component {
    static propTypes = {
        id: PropTypes.string.isRequired
    } 

    update = (cache, payload) => {
        const data = cache.readQuery({query: CURRENT_USER_QUERY});

        // this is basically the specified payload from your mutation
        const cartItemId = payload.data.removeFromCart.id

        data.me.cart = data.me.cart.filter(cartItem => cartItem.id !== cartItemId);

        cache.writeQuery({query: CURRENT_USER_QUERY, data});
    }

    render() {
        return (
            // prisma automatically injects typenames, but since we're doing this
            // on the client, we need to add them ourselves
            <Mutation mutation={REMOVE_FROM_CART_MUTATION} 
                    update={this.update}
                    optimisticResponse={{
                        __typename: 'Mutation',
                        removeFromCart: {
                            __typename: 'CartItem',
                            id: this.props.id
                        }
                    }}>
                { (removeFromCart, {loading, error}) => (
                    <BigButton
                        disabled={loading}
                        onClick={()=>{
                            // you can also pass the variables into this function if 
                            // you prefer (I personally like how that looks)
                            return removeFromCart({variables: { id: this.props.id}}).catch( err => console.log(err.message));
                        } }
                        title="Delete Item" > &times;</BigButton>
                ) }
            </Mutation>
        );
    }
}

export default RemoveFromCart;
export {REMOVE_FROM_CART_MUTATION}; 