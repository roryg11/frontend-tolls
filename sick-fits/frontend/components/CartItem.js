import styled from "styled-components";
import formatMoney from "../lib/formatMoney";
import RemoveFromCart from "./RemoveFromCart";

const CartStyles = styled.li`
    padding: 1rem 0;
    display: grid;
    grid-template-columns: auto 1fr auto;
    border-bottom: 1px solid ${ props => props.theme.lightGrey};
    align-items: center;
    img {
        margin-right: 10px;
    }
    h3, p {
        margin: 0; 
    }
`

const CartItem = props => {
    const {cartItem} = props;
    if(!cartItem.item){ 
        return (
            <CartStyles>
                <div>&nbsp;</div>
                <p>This item has been removed</p>
                <RemoveFromCart id={cartItem.id}/>
            </CartStyles>
        )
    };
    return (
        <CartStyles>
            <img width="100" src={cartItem.item.image}/>
            <div className="cart-item-details">
                <h3>{cartItem.item.title}</h3>
                <p>
                    {formatMoney(cartItem.item.price * cartItem.quantity)}
                    {' - '}
                    <em>
                        {cartItem.quantity} &times; {formatMoney(cartItem.item.price)} each
                    </em>
                </p>
            </div>
            <RemoveFromCart id={cartItem.id}/>
        </CartStyles>
    )
}

export default CartItem; 