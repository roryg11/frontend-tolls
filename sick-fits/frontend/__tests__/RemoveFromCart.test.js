import {mount} from "enzyme";
import toJSON from "enzyme-to-json";
import { ApolloConsumer } from "react-apollo";
import { MockedProvider } from "react-apollo/test-utils";
import wait from "waait";
import RemoveFromCart, { REMOVE_FROM_CART_MUTATION }  from "../components/RemoveFromCart";
import {CURRENT_USER_QUERY} from "../components/User";
import { fakeUser, fakeCartItem, fakeItem } from "../lib/testUtils";

const user = fakeUser();
const cartItem = fakeCartItem();


const mocks = [
    {
        request: { query: CURRENT_USER_QUERY },
        result: {
            data: {
                me: {
                    ...user,
                    cart: [cartItem]
                },
                
            }
        }
    },
    {
        request: { query: REMOVE_FROM_CART_MUTATION, variables: {id: cartItem.id}},
        result: {
            data: {
                removeFromCart: {
                    id: cartItem.id
                }
            } 
        }
    }
];

describe('<RemoveFromCart/>', ()=>{
    it('renders and matches snapshot', ()=>{
        const wrapper = mount(
            <MockedProvider>
                <RemoveFromCart id={cartItem.id}/>
            </MockedProvider>
        );

        expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('removes item from users cart', async ()=>{
        let apolloClient;
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <ApolloConsumer>
                    { (client)=>{
                        apolloClient = client;
                        return(
                            <RemoveFromCart id={cartItem.id}/>
                        )
                    }}
                </ApolloConsumer>
            </MockedProvider>
        );

        await wait();
        wrapper.update();

        const {data: {me: currentUser}} = await apolloClient.query({query: CURRENT_USER_QUERY});
        expect(currentUser.cart.length).toEqual(1);

        wrapper.find('button').simulate('click');

        await wait();
        wrapper.update();

        const {data: {me: updatedUser}} = await apolloClient.query({query: CURRENT_USER_QUERY});
        expect(updatedUser.cart.length).toEqual(0);
    });
});