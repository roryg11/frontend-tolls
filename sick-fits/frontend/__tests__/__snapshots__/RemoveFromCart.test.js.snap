// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RemoveFromCart/> renders and matches snapshot 1`] = `
<MockedProvider
  addTypename={true}
>
  <ApolloProvider
    client={
      ApolloClient {
        "__operations_cache__": Map {},
        "cache": InMemoryCache {
          "addTypename": true,
          "config": Object {
            "addTypename": true,
            "dataIdFromObject": [Function],
            "fragmentMatcher": HeuristicFragmentMatcher {},
            "storeFactory": [Function],
          },
          "data": ObjectCache {
            "data": Object {},
          },
          "optimistic": Array [],
          "silenceBroadcast": false,
          "typenameDocumentCache": WeakMap {},
          "watches": Array [],
        },
        "defaultOptions": Object {},
        "disableNetworkFetches": false,
        "link": ApolloLink {
          "request": [Function],
        },
        "mutate": [Function],
        "query": [Function],
        "queryDeduplication": true,
        "reFetchObservableQueries": [Function],
        "resetStore": [Function],
        "resetStoreCallbacks": Array [],
        "ssrMode": false,
        "store": DataStore {
          "cache": InMemoryCache {
            "addTypename": true,
            "config": Object {
              "addTypename": true,
              "dataIdFromObject": [Function],
              "fragmentMatcher": HeuristicFragmentMatcher {},
              "storeFactory": [Function],
            },
            "data": ObjectCache {
              "data": Object {},
            },
            "optimistic": Array [],
            "silenceBroadcast": false,
            "typenameDocumentCache": WeakMap {},
            "watches": Array [],
          },
        },
        "version": "2.4.2",
        "watchQuery": [Function],
      }
    }
  >
    <RemoveFromCart
      id="omg123"
    >
      <Mutation
        mutation={
          Object {
            "definitions": Array [
              Object {
                "directives": Array [],
                "kind": "OperationDefinition",
                "name": Object {
                  "kind": "Name",
                  "value": "REMOVE_FROM_CART_MUTATION",
                },
                "operation": "mutation",
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "id",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "id",
                            },
                          },
                        },
                      ],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "removeFromCart",
                      },
                      "selectionSet": Object {
                        "kind": "SelectionSet",
                        "selections": Array [
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "id",
                            },
                            "selectionSet": undefined,
                          },
                        ],
                      },
                    },
                  ],
                },
                "variableDefinitions": Array [
                  Object {
                    "defaultValue": undefined,
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "ID",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                    },
                  },
                ],
              },
            ],
            "kind": "Document",
            "loc": Object {
              "end": 115,
              "start": 0,
            },
          }
        }
        optimisticResponse={
          Object {
            "__typename": "Mutation",
            "removeFromCart": Object {
              "__typename": "CartItem",
              "id": "omg123",
            },
          }
        }
        update={[Function]}
      >
        <RemoveFromCart__BigButton
          disabled={false}
          onClick={[Function]}
          title="Delete Item"
        >
          <button
            className="RemoveFromCart__BigButton-emvtd6-0 gPWNSn"
            disabled={false}
            onClick={[Function]}
            title="Delete Item"
          >
             Ã—
          </button>
        </RemoveFromCart__BigButton>
      </Mutation>
    </RemoveFromCart>
  </ApolloProvider>
</MockedProvider>
`;
